File: bindings/dotnet/ChatModel.cs
using System;
using System.ComponentModel;
using System.Threading;
using System.Threading.Tasks;
using Arut.Ffi;

namespace Arut.Bindings;

public sealed class ChatModel : INotifyPropertyChanged, IDisposable
{
    private readonly ProductSessionHandle session;
    private readonly CancellationTokenSource lifetime = new();
    private ChatHandle core;
    private ComposerHandle composer;
    private ObservableState<ChatState> chatState;
    private ObservableState<ComposerState> composerState;
    private Task draftWrites = Task.CompletedTask;
    private string draft = "";
    private long epoch;

    public ChatModel()
    {
        session = Arut_ffi.CreateProductSession("", "local-demo");
        core = session.Chat();
        composer = core.Composer();
        chatState = new(core.State(), core.State, core.ChatChanges);
        composerState = new(composer.State(), composer.State, composer.ComposerChanges);
        chatState.Changed += ChatChanged;
        composerState.Changed += ComposerChanged;
        _ = InitializeAndPoll(Volatile.Read(ref epoch), lifetime.Token);
    }

    public ChatState State => chatState.Value;
    public ComposerState Composer => composerState.Value;
    public ChatSummary[] History => session.ChatSummaries();
    public string SelectedChatId => core.Id();

    public string Draft
    {
        get => draft;
        set
        {
            if (draft == value) return;
            draft = value;
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Draft)));
            var expectedEpoch = Volatile.Read(ref epoch);
            draftWrites = WriteDraftAfter(draftWrites, composer, value, expectedEpoch);
        }
    }

    public async Task Send()
    {
        var expectedEpoch = Volatile.Read(ref epoch);
        var targetComposer = composer;
        var targetChat = core;
        await draftWrites;
        var message = draft.Trim();
        if (message.Length == 0 || State.Status == ChatStatus.Sending || expectedEpoch != Volatile.Read(ref epoch)) return;
        await targetChat.Send(message);
        if (expectedEpoch == Volatile.Read(ref epoch))
        {
            ApplyComposer(targetComposer.State());
            PublishChatProperties();
        }
    }

    public void NewChat() => Bind(session.NewChat());

    public void SelectChat(string chatId) => Bind(session.SelectChat(chatId));

    public string[] ChatIds() => session.ChatIds();

    public void Dispose()
    {
        lifetime.Cancel();
        Interlocked.Increment(ref epoch);
        chatState.Dispose();
        composerState.Dispose();
        composer.Dispose();
        core.Dispose();
        session.Dispose();
        lifetime.Dispose();
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    private void Bind(ChatHandle next)
    {
        Interlocked.Increment(ref epoch);
        chatState.Changed -= ChatChanged;
        composerState.Changed -= ComposerChanged;
        chatState.Dispose();
        composerState.Dispose();
        composer.Dispose();
        core.Dispose();
        core = next;
        composer = core.Composer();
        chatState = new(core.State(), core.State, core.ChatChanges);
        composerState = new(composer.State(), composer.State, composer.ComposerChanges);
        chatState.Changed += ChatChanged;
        composerState.Changed += ComposerChanged;
        ApplyComposer(composer.State());
        ChatChanged();
        _ = InitializeAndPoll(Volatile.Read(ref epoch), lifetime.Token);
    }

    private async Task WriteDraftAfter(Task previous, ComposerHandle target, string text, long expectedEpoch)
    {
        await previous;
        if (expectedEpoch != Volatile.Read(ref epoch)) return;
        await target.Replace(text);
    }

    private async Task InitializeAndPoll(long expectedEpoch, CancellationToken cancellationToken)
    {
        var target = composer;
        try
        {
            await target.Initialize(cancellationToken);
            while (!cancellationToken.IsCancellationRequested && expectedEpoch == Volatile.Read(ref epoch))
            {
                await Task.Delay(500, cancellationToken);
                if (expectedEpoch != Volatile.Read(ref epoch)) return;
                await draftWrites;
                var state = await target.SyncOnce(cancellationToken);
                if (expectedEpoch == Volatile.Read(ref epoch)) ApplyComposer(state);
            }
        }
        catch (OperationCanceledException) { }
    }

    private void ChatChanged() =>
        PublishChatProperties();

    private void PublishChatProperties()
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(State)));
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(History)));
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(SelectedChatId)));
    }

    private void ComposerChanged() => ApplyComposer(composerState.Value);

    private void ApplyComposer(ComposerState state)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Composer)));
        if (!draftWrites.IsCompleted) return;
        if (draft == state.Text) return;
        draft = state.Text;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Draft)));
    }
}

File: bindings/swift/Sources/ArutBindings/ChatModel.swift
import ArutFfi
import Combine
import Foundation

@MainActor
public final class ChatModel: ObservableState<ChatState> {
    private let session: ProductSessionHandle
    private var core: ChatHandle
    private var composer: ComposerHandle
    private var cancelComposer: (() -> Void)?
    private var polling: Task<Void, Never>?
    private var epoch: UInt64 = 0
    private var applyingDraft = false
    private var pendingEdits = 0
    @Published public var draft = "" {
        didSet {
            guard !applyingDraft, draft != oldValue else { return }
            let text = draft
            let expectedEpoch = epoch
            let target = composer
            pendingEdits += 1
            Task {
                defer { pendingEdits -= 1 }
                guard expectedEpoch == epoch else { return }
                if let state = try? await target.replace(text: text), pendingEdits == 1 {
                    apply(state)
                }
            }
        }
    }

    public init() {
        let session = createProductSession(backendUrl: "", pendingScopeId: "local-demo")
        let core = session.chat()
        self.session = session
        self.core = core
        self.composer = core.composer()
        super.init(
            read: core.state,
            subscribe: { invalidate in
                let changes = core.chatChanges(callback: invalidate)
                return changes.cancel
            }
        )
        bindComposer()
    }

    public func send() {
        let message = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !message.isEmpty, state.status != .sending else { return }
        let expectedEpoch = epoch
        let targetComposer = composer
        let targetChat = core
        Task {
            _ = try? await targetComposer.replace(text: message)
            guard expectedEpoch == epoch else { return }
            _ = try? await targetChat.send(text: message)
            guard expectedEpoch == epoch else { return }
            apply(targetComposer.state())
            receive(targetChat.state())
        }
    }

    public func newChat() {
        bind(session.newChat())
    }

    public var chatIDs: [String] {
        session.chatIds()
    }

    public var history: [ChatSummary] {
        session.chatSummaries()
    }

    public var selectedChatID: String {
        core.id()
    }

    public func selectChat(_ chatID: String) {
        bind(session.selectChat(chatId: chatID))
    }

    private func bind(_ next: ChatHandle) {
        epoch &+= 1
        polling?.cancel()
        cancelComposer?()
        core = next
        composer = next.composer()
        receive(next.state())
        observe(
            read: next.state,
            subscribe: { invalidate in
                let changes = next.chatChanges(callback: invalidate)
                return changes.cancel
            }
        )
        bindComposer()
    }

    private func bindComposer() {
        let expectedEpoch = epoch
        let target = composer
        apply(target.state())
        cancelComposer = target.composerChanges { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self, self.epoch == expectedEpoch else { return }
                self.apply(target.state())
            }
        }.cancel
        polling = Task { [weak self] in
            guard let self else { return }
            _ = try? await target.initialize()
            while !Task.isCancelled, self.epoch == expectedEpoch {
                try? await Task.sleep(nanoseconds: 500_000_000)
                guard !Task.isCancelled, self.epoch == expectedEpoch else { return }
                if let state = try? await target.syncOnce() {
                    self.apply(state)
                }
            }
        }
    }

    private func apply(_ state: ComposerState) {
        guard pendingEdits == 0 else { return }
        guard draft != state.text else { return }
        applyingDraft = true
        draft = state.text
        applyingDraft = false
    }

    deinit {
        polling?.cancel()
        cancelComposer?()
        stopObserving()
    }
}
